svcdiscovery/plugin/svcdiscovery/plugin.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2020 SUSE
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package svcdiscovery
+
+import (
+	"context"
+	"net"
+
+	"github.com/coredns/coredns/plugin"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+)
+
+// ServiceDiscovery is a CoreDNS plugin for handling Cloud Foundry App Service
+// Discovery.
+type ServiceDiscovery struct {
+	Next plugin.Handler
+
+	log       clog.P
+	sdcClient *SDCClient
+	ttl       uint32
+}
+
+// Name returns the plugin name.
+func (sd *ServiceDiscovery) Name() string { return pluginName }
+
+// ServeDNS handles domain name requests for Cloud Foundry App Service Discovery
+// by resolving internal domain names.
+func (sd *ServiceDiscovery) ServeDNS(
+	ctx context.Context,
+	res dns.ResponseWriter,
+	req *dns.Msg,
+) (int, error) {
+	state := request.Request{W: res, Req: req}
+	qclass := state.QClass()
+	qtype := state.QType()
+	qname := state.Name()
+
+	if qclass == dns.ClassINET && (qtype == dns.TypeA || qtype == dns.TypeAAAA) {
+		ips, err := sd.sdcClient.Discover(qname)
+		if err != nil {
+			sd.log.Error(err)
+			return dns.RcodeServerFailure, err
+		}
+
+		if len(ips) > 0 {
+			res = &ServiceDiscoveryResponseWriter{
```

Comment:
Hmm, their sample for a plugin that actually resolves a thing is the [whoami](https://github.com/coredns/coredns/blob/master/plugin/whoami/whoami.go) plugin.  That code makes more sense: if it's got a proper response, just use the `ResponseWriter` to write it out, and don't pass it onto the next plugin.  That would mean we don't need to write the whole `ResponseWriter`.

In our case though we probably want to chain to the next plugin if we would have `NXDOMAIN`ed.

Reply:
Fixed with ac0eb20.